pkg/config/generator.go

```diff
@@ -70,6 +73,39 @@ func NewMultusConfig(pluginName string, cniVersion string, kubeconfig string, co
 // Generate generates the multus configuration from whatever state is currently
 // held
 func (mc *MultusConf) Generate() (string, error) {
+	versionFmt := "delegate cni version is %s while top level cni version is %s"
+	v040, _ := semver.Make("0.4.0")
+	v031, _ := semver.Make("0.3.1")
+	topLevel, err := semver.Make(mc.CNIVersion)
+
+	if err != nil {
+		s := "Error in top level cni version"
+		return "", errors.New(s)
+	}
+
+	if topLevel.GTE(v040) {
+		for _, delegate := range mc.Delegates {
+			delegatesMap, ok := delegate.(map[string]interface{})
+			if !ok {
+				s := "Couldn't get cni version of delegate"
+				return "", errors.New(s)
+			}
+			delegateVersion, ok := delegatesMap["cniVersion"].(string)
+			if !ok {
+				s := "Couldn't get cni version of delegate"
+				return "", errors.New(s)
+			}
+			v, err := semver.Make(delegateVersion)
+			if err != nil {
+				return "", err
+			}
+			if v.LTE(v031) {
```

Comment:
I don't understand why 0.3.1 is special here. Shouldn't we just reject *any* delegates whose version is lower than 0.4.0 ?

Reply:
I addressed this by only checking nested version < 0.4.0 when top level version >= 0.4.0.